### PR TITLE
First Unit Tests Online

### PR DESCRIPTION
**Pull request by xforever1313**
This pull request puts the first unit tests online.  It also fixes a few potential bugs in the configuration validation:

- Fixed an issue where if sendStickerEmoji, showJoinMessage. showLeaveMessage were NOT in the settings, were undefined, or null, we would crash.  We now set these to default values instead of crashing.

- Added an TeleIrcException class, along with an Error code class.

- Added default settings to the TeleIrc Class.

- Added unit tests for the IrcConfig validation.  Unit tests ensure that if required settings are missing, are null, or are undefined, we throw Exceptions.  Anything that is option is set to the default if they are null, undefined, or missing.

To run unit tests, install nodeunit, and on the command line run in the tests directory "nodeunit ./".

### CODE REVIEW

File: tests/IrcConfigValidationTests.js [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+            botName: sampleSettings.irc.botName
+        },
+        ircBlacklist: sampleSettings.ircBlacklist,
+        tg: sampleSettings.tg
+    }
+
+    doDefaultSettingsTest(assert, testSettings);
+
+    assert.done();
+};
```
Comment (thenaterhood) [inline tests/IrcConfigValidationTests.js]: Missing newline at the end of the file

File: tests/IrcConfigValidationTests.js [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+
+/**
+ * Ensures if the IRC config's optional values are set to undefined, it gets
+ * set to its default.
+ */
+exports.ircConfigValidation_defaultSettingsTest_missingSettings= function(assert) {
```
Comment (thenaterhood) [inline tests/IrcConfigValidationTests.js]: There should be a space before the `=` per language convention

File: tests/IrcConfigValidationTests.js [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+    doDefaultSettingsTest(assert, testSettings);
+
+    assert.done();
+};
+
+/**
+ * Ensures if the IRC config is missing optional values, it gets
+ * added in as a default.
+ */
+exports.ircConfigValidation_defaultSettingsTest_setToNull= function(assert) {
```
Comment (thenaterhood) [inline tests/IrcConfigValidationTests.js]: There should be a space before the `=` per language convention

File: tests/IrcConfigValidationTests.js [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+    assert.strictEqual(uut.defaultIrcSuffix, testSettings.irc.suffix);
+    assert.strictEqual(uut.defaultShowJoinMessage, testSettings.irc.showJoinMessage);
+    assert.strictEqual(uut.defaultShowLeaveMessage, testSettings.irc.showLeaveMessage);
+}
+
+/**
+ * Ensures if the IRC config is missing optional values, it gets
+ * added in as a default.
+ */
+exports.ircConfigValidation_defaultSettingsTest_notDefined= function(assert) {
```
Comment (thenaterhood) [inline tests/IrcConfigValidationTests.js]: There should be a space before the `=` per language convention

File: tests/IrcConfigValidationTests.js [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+    doRequiredSettingsTest(assert, uut, uut.errorCodes.MissingIrcChannelConfig);
+
+    assert.done();
+}
+
+/**
+ * Ensures if the IRC config is missing its channel config, an
+ * exception gets thrown.
+ */
+exports.ircConfigValidation_missingBotNameConfig= function(assert) {
```
Comment (thenaterhood) [inline tests/IrcConfigValidationTests.js]: There should be a space before the `=` per language convention

File: tests/IrcConfigValidationTests.js [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+    doRequiredSettingsTest(assert, uut, uut.errorCodes.MissingIrcServerConfig);
+
+    assert.done();
+}
+
+/**
+ * Ensures if the IRC config is missing its channel config, an
+ * exception gets thrown.
+ */
+exports.ircConfigValidation_missingChannelConfig= function(assert) {
```
Comment (thenaterhood) [inline tests/IrcConfigValidationTests.js]: There should be a space before the `=` per language convention

File: tests/IrcConfigValidationTests.js [hunk 7]
```diff
@@ -0,0 +1,312 @@
...
+        uut.initStage1_ircConfigValidation.bind(uut),
+        expectedErrorCode
+    );
+}
+
+/**
+ * Ensures if the IRC config is missing its server config, an
+ * exception gets thrown.
+ */
+exports.ircConfigValidation_missingServerConfig= function(assert) {
```
Comment (thenaterhood) [inline tests/IrcConfigValidationTests.js]: There should be a space before the `=` per language convention

File: tests/IrcConfigValidationTests.js [hunk 8]
```diff
@@ -0,0 +1,312 @@
...
+        }
+    );
+}
+
+// -------- Missing IRC Config Tests --------
+
+/**
+ * Ensures if the IRC config is missing completely from our settings, an exception gets thrown.
+ */
+exports.ircConfigValidation_MissingIrcConfig= function(assert) {
```
Comment (thenaterhood) [inline tests/IrcConfigValidationTests.js]: There should be a space before the `=` per language convention

File: lib/libteleirc.js [hunk 9]
```diff
@@ -1,5 +1,58 @@
...
+class TeleIrcErrorCodes {
+    constructor() {
+        // Let 1XXX be irc configuration errors.
+
+        this.MissingIrcConfig = 1000; // IRC config is missing from the config, is undefined, or is null.
+        this.MissingIrcServerConfig = 1001; // IRC config is missing its server information.
+        this.MissingIrcChannelConfig = 1002; // IRC config is missing its channel information.
+        this.MissingIrcBotNameConfig = 1003; // IRC config is missing its bot name.
+
+        // Let 2XXX be Telegram Configuration errors.
```
Comment (thenaterhood) [inline lib/libteleirc.js]: Is this for future work, or is there supposed to be something here?

**Comment by jflory7** [comment]
@xforever1313 Think you'll have a chance to swing by and amend the PR?

**Comment by xforever1313** [comment]
@jflory7, I can work on things tomorrow night.   Been sick with a cold, and still down about the election 😳.

**Comment by jflory7** [comment]
@xforever1313 No worries, not a rush for this – just wanted to check up on it. Some time off to process is more than justified.

**Comment by xforever1313** [comment]
Ready for Review.

**Comment by thenaterhood** [comment]
Looks fine. just out of date.

**Comment by robbyoconnor** [comment]
You'll need to rebase and perhaps update tests -- since I added [dotenv](https://www.npmjs.com/package/dotenv) in #26

**Comment by robbyoconnor** [comment]
You'll need to rebase and perhaps update tests -- since I added [dotenv](https://www.npmjs.com/package/dotenv) in #26 -- docker-compose uses the same file dotenv uses for setting environment vars.

**Comment by xforever1313** [comment]
Repulled as #28, since conflicts are hard :|.

**Pull request closed without merging**